resources/js/Events/Components/Core/EventTree.jsx

```diff
@@ -0,0 +1,281 @@
+          innerFlatTree[0].parent_id = root.id;
+          childs.push(...innerFlatTree);
+        });
+
+        return [root, ...childs];
+      }
+
+      if (eventData.hasOwnProperty("data")) {
+        const root = {
+          label: eventData.name,
+          path: `${parentPath}>>${eventData.name}`,
+          id: `${parentPath}>>${eventData.name}`,
+          state: "unchecked",
+          expand: level == 0,
+          children: []
+        };
+
+        const childs = [];
+
+        eventData.data.forEach((e) => {
+          let eventNode = {
+            label: e.short_label ?? e.label,
+            path: `${root.path}>>${e.short_label ?? e.label}`,
+            id: `${root.path}>>${e.short_label ?? e.label}`,
+            state: "unchecked",
+            expand: level == 0,
+            data: e
+          };
+
+          root.children.push(eventNode.id);
+          eventNode.parent_id = root.id;
+
+          childs.push(eventNode);
+        });
+
+        return [root, ...childs];
+      }
+    };
+
+    const root = {
+      label: source,
+      path: source,
+      id: source,
+      state: "unchecked",
+      parent_id: null,
+      expand: true,
+      children: []
+    };
+
+    const childs = [];
+
+    events.forEach((ed) => {
+      let innerFlatTree = makeFlatTreeInner(ed, source);
+      root.children.push(innerFlatTree[0].id);
+      innerFlatTree[0].parent_id = root.id;
+      childs.push(...innerFlatTree);
+    });
+
+    const res = {};
+
+    [root, ...childs].forEach((e) => {
+      res[e.id] = e;
+    });
+
+    return new EventTree(res);
+  }
+
+  toggleCheckbox(id) {
+    const stateCycle = {
+      checked: "unchecked",
+      unchecked: "checked",
+      indecided: "checked"
+    };
+
+    const newState = stateCycle[this[id]["state"]];
+
+    const nodes = this.getNodes(id);
+
+    for (const n of nodes) {
+      this[n].state = newState;
+    }
+
+    let parentId = this[id].parent_id;
+
+    while (parentId != null) {
+      let siblingsStates = this[parentId].children.map((cid) => this[cid].state);
+
+      let parentState = "indecided";
+
+      if (siblingsStates.every((ss) => ss == "checked")) {
+        parentState = "checked";
+      }
+
+      if (siblingsStates.every((ss) => ss == "unchecked")) {
+        parentState = "unchecked";
+      }
+
+      this[parentId].state = parentState;
+
+      parentId = this[parentId].parent_id;
+    }
+
+    return new EventTree({ ...this });
+  }
+
+  toggleExpand(id) {
+    if (this.hasOwnProperty(id)) {
+      this[id].expand = !this[id].expand;
+    }
+
+    return new EventTree({ ...this });
+  }
+
+  getNodes(id) {
+    const isEvent = this[id].hasOwnProperty("data");
+
+    if (isEvent) {
+      return [id];
+    }
+
+    const allNodes = [id];
+
+    this[id].children.forEach((cid) => {
+      allNodes.push(...this.getNodes(cid));
+    });
+
+    return allNodes;
+  }
+
+  applySelections(selections) {
+    selections.forEach((s) => {
```

Comment:
please rename `s` to `id` in this function
```suggestion
    selections.forEach((id) => {
```